### ISSUE TITLE
Let binding pattern matching to unpack records

### ISSUE THREAD

**Issue by ocharles** [dhall-lang/dhall-lang#74]
The ability to unpack a record through a `let`-binding is tremendously powerful for building a lightweight module system. I could envision doing something like:

```dhall
let { map, filter, span } = http://prelude.dhall/List
in ...
```

Right now, my only option is to import each function individually. Not only is that more verbose, it makes it much harder to ensure consistency. While the above functions are separate, one could imagine an import where we want to make sure that both function comes from the same "version" of a library, and the only make sense (operationally) if we're using the same version.

**Comment by Gabriel439** [comment]
There's a correlated issue: currently Dhall records don't support types as fields (which you would probably want if you used this for a module system), but that is an old restriction that I might be able to lift now that Dhall supports type synonyms.  I just need to do some additional work to verify that Dhall is sound with generalized records.

I think this is a good idea, but I just need to think a little bit more about what other changes to the language that it would entail.  For example, this overlaps substantially with the `let inherit` feature in Nix:

```nix
let inherit (someRecord) map filter span;
in  ...
```

... which makes me suspect that if we add this then people may also begin to request support for `inherit`-like functionality for Dhall records (which I also think is probably a good idea), but if we do that then I would like the syntax to be consistent between the two like in Nix.

**Comment by Gabriel439** [comment]
Another thing to consider: if we can pattern match in a `let`-binding, should we be able to do so in an anonymous function, too?

For example:

```haskell
λ({ map, filter, span } : …) → …
```

In fact, there are a lot of other places in the grammar that use `label` that might benefit from the ability to replace a label with a set of labels, such as:

* Accessing multiple fields.  For example `record.{map, filter, span}` might project out just those fields
* Defining multiple fields of record literals of record types.  For example `{ {x, y, z} = 2 }` might set all three fields with the same value
* Defining multiple types of union alternatives.  For example `< {x, y, z} : {} >` would be a convenient short-hand syntax for defining an enum

**Comment by Gabriel439** [comment]
The other reason I brought up anonymous functions specifically is that they would play nice with `merge` expressions to more closely resemble Haskell-style pattern matching:

```
merge
{ Constructor1 = λ({ field1A, field1B } : …) → …
, Constructor2 = λ({ field2A } : …) → …
}
…
```

**Comment by ocharles** [comment]
The projection is a pretty cool idea that I could see having use in https://github.com/ocharles/dhallql, if I ever pick that project up again.

For the others I'm less opinionated on. I thought it might be best to just do this to `let` as it keeps the language change small and we can always add in the extra pattern matching later. It's much harder to take it out if it's not a good feature.

**Comment by Gabriel439** [comment]
Yeah, to clarify, we don't have to implement all of those features in one go.  But I'd like to do some look-ahead design so that if we do add those features later on we aren't painting ourselves into a corner with a bad initial choice of syntax

**Comment by Profpatsch** [comment]
I would also add an exhaustiveness check if `..` is not used (see https://discourse.dhall-lang.org/t/exhaustive-record-field-matching/75/1 for motivation).

e.g.
* `let { foo } = { foo = "abc", bar = "cde" }` => error, field `bar` exists but was not matched
* `let { foo, ... } = { foo = "abc", bar = "cde" }` => no error, partial matching was requested

This is particularly helpful if dhall detects unused bindings and warns accordingly.

**Comment by sjakobi** [comment]
If we have exhaustiveness checking, I believe we'll also need a way of expressing "I know this field exists but I don't want to use it" without opting out of the exhaustive check completely via `..`.

Maybe this syntax could work for that:

```dhall
let { foo, bar, (baz) } = { foo = 1, bar = 2, baz = "bla" }

in foo + bar
```

**Comment by Nadrieril** [comment]
More generally we may want to rename the fields; this would make it closer to a form of pattern-matching. We'd get something like that, similar to what Rust does:
```rust
let { foo, bar: renamed_bar, baz: _ } = { foo = 1, bar = 2, baz = "bla" }
in foo + renamed_bar
```
Where the `_` is just a binding to a variable named `_`, but by convention this indicates that the variable is not used, and would replace you `(baz)` suggestion.

**Comment by sjakobi** [comment]
@Nadrieril Ah, yes, that's better!

`dhall lint` should also detect unused bindings here, and turn e.g.

```
let { foo, bar } = { foo = 1, bar = 2 }

in foo
```

into

```
let { foo, bar: _ } = { foo = 1, bar = 2 }

in foo
```

**Comment by ocharles** [comment]
I would much rather a keyword like 'as' or 'renaming', than using ':'

**Comment by sjakobi** [comment]
```dhall
let { foo, bar as renamed_bar, baz as _ } = { foo = 1, bar = 2, baz = "bla" }

in foo + renamed_bar
```

seems reasonable to me, although I don't have much of a preference.

I'm actually not a huge fan of the

    baz as _

bit. I think I'd slightly prefer

```dhall
let { foo, bar as renamed_bar, (baz) } = { foo = 1, bar = 2, baz = "bla" }

in foo + renamed_bar
```

**Comment by ocharles** [comment]
I would rather '...' there, to say "and other stuff"

**Comment by sjakobi** [comment]
@ocharles The problem with `..` or `...` is that it opts out of the exhaustiveness check, which is reasonable for imports, but not for some other uses.

**Comment by SiriusStarr** [comment]
I think that the ability to opt-out of the exhaustiveness check is good, as long as it's explicitly opt-**out**.  I would like the ability to write both functions that I am **forced** to update when a record changes (e.g. the `printRecord` function you mentioned @sjakobi in the other issue) and functions that I don't have to update (e.g. `printXYCoordinates` shouldn't care what other fields are in the record).  (Obviously, extensible records is what I'd really like for the latter, but this is a step in that direction, haha.)

Using `(field)` for opting-out of unpacking a field seems possibly confusing, just since parentheses are used pretty extensively to mean something else (grouping) elsewhere in the grammar.  I think I'd prefer the `baz as _` paradigm (or the use of some other delimiter without a well-defined meaning).

Is there general consensus for (what probably isn't, but I'm going to call) Elm-style renaming, e.g. `bar as renamed_bar` over Haskell-style `renamed_bar@bar`?  (I probably lean slightly towards the former but don't have a strong preference.)

Another question, is nested unpacking going to be allowed, e.g. can you do:
```
let r = { foo : Natural, bar : {x : Natural, y : Natural} }

let { foo, bar as { x, y } } = r
```
This is not allowed by #863 but seems like it potentially should be here.

**Comment by sjakobi** [comment]
What I don't like so much about `baz as _` is that `_` is just a regular old variable name in other contexts where it doesn't mean "ignore me".


> Is there general consensus for (what probably isn't, but I'm going to call) Elm-style renaming, e.g. `bar as renamed_bar` over Haskell-style `renamed_bar@bar`? (I probably lean slightly towards the former but don't have a strong preference.)

`@` is used for `DeBruijn`-indices when referencing variables, but maybe that's not so bad. This looks ok to me:

```dhall
let { foo, renamed_bar@bar } = { foo = 1, bar = 2 }

in foo + renamed_bar
```

> Another question, is nested unpacking going to be allowed, e.g. can you do:
> 
> ```
> let r = { foo : Natural, bar : {x : Natural, y : Natural} }
> 
> let { foo, bar as { x, y } } = r
> ```
> 
> This is not allowed by #863 but seems like it potentially should be here.

I think this is kind of neat, but would add even more complexity to a not so simple proposal, so I think I'm rather -1 on that. Also, this doesn't look too bad to me.

```
let r = { foo : Natural, bar : {x : Natural, y : Natural} }

let { foo, bar } = r

let {x, y} = bar

in ...
```

Maybe we could consider adding nested destructuring later on…

**Comment by sjakobi** [comment]
> Another question, is nested unpacking going to be allowed, e.g. can you do:
> 
> ```
> let r = { foo : Natural, bar : {x : Natural, y : Natural} }
> 
> let { foo, bar as { x, y } } = r
> ```

Also, to be clear: So far we've discussed destructuring record _values_ here. I don't think we need this for record _types_ right now.

**Comment by sjakobi** [comment]
And more clarification. I want to have _both_ `..` or `...` for opting out of the exhaustiveness check _and_ a way of ignoring fields while _not_ opting out of the exhaustiveness. So something like this should be valid:

```
let r = { foo = 1, bar = 2 }

let { foo, .. } = r

let { bar, (foo) } = r

in ...
```

**Comment by ocharles** [comment]
WRT to @ and as, my real preference is to reduce overloading. @ is taken,
and I see no really compelling reason to further overload what it means

**Comment by SiriusStarr** [comment]
> What I don't like so much about baz as _ is that _ is just a regular old variable name in other contexts where it doesn't mean "ignore me".

Personally, I would like to have `_` attain its "usual" (Haskell et al.) status; it seems a pretty common convention, and I doubt there is much in the way of extant Dhall codebases that use `_` alone as a variable name other than for the purposes of ignoring it.  (I certainly use it when I need to write a `const` equivalent or whatnot in Dhall.)

Failing that, one could possibly overload the exhaustiveness check here and do `{foo, bar as .. , baz}`; should be unambiguous to parse, and whatever that operator ends up being is going to have to become a reserved identifier anyways in the course of this proposal (so that `{foo, bar, .. }` doesn't get parsed as 3 field names).

> Also, to be clear: So far we've discussed destructuring record _values_ here. I don't think we need this for record _types_ right now.

For like the fourth time, my typos feel personally attacked, @sjakobi .  ;-)  (I obviously didn't mean to do that to a record type, was just trying to write something quickly as I was running out the door, haha.  Edited to correct it.)

>@ is already used to separate the DeBruijn-indices when referencing variables

> WRT to @ and as, my real preference is to reduce overloading. @ is taken,

Derp, I totally forgot that `@` is used; I absolutely agree it's not a good choice given that.

> I think this [nested destructuring] is kind of neat, but it would add even more complexity to a not so simple proposal

I agree that it's not something I would put a ton of work into making happen at this point, I just am thinking it may naturally shake out in the grammar to be a fairly easy addition.

**Comment by sjakobi** [comment]
> Personally, I would like to have `_` attain its "usual" (Haskell et al.) status; it seems a pretty common convention, and I doubt there is much in the way of extant Dhall codebases that use `_` alone as a variable name other than for the purposes of ignoring it. (I certainly use it when I need to write a `const` equivalent or whatnot in Dhall.)

`_` is in fact the _only_ variable name you'll find in α-normalized Dhall, which ultimately is just plain old Dhall.

But this is not very important to me. If everyone prefers `baz as _` over `(baz)`, I won't protest.

**Comment by SiriusStarr** [comment]
> _ is in fact the only variable name you'll find in α-normalized Dhall, which ultimately is just plain old Dhall.

Okay, fair.  XD

> But this is not very important to me. If everyone prefers baz as _ over (baz), I won't protest.

I'm good with something along the lines of `(baz)`, I just don't think it should be parentheses given that they have such a strongly established meaning.  Thoughts on using a different delimiter?  `_baz_`?  `~baz~`?  `-baz-`?

**Comment by Nadrieril** [comment]
I agree that the `{ bar: renamed_bar }` syntax is confusing. What about `{ bar = renamed_bar }`, so that deconstruction mirrors construction ?
My secret plan is that we eventually get proper pattern-matching (instead of `merge`), and then we'll want to have deconstruction and construction look similar more generally.

**Comment by SiriusStarr** [comment]
> What about { bar = renamed_bar }, so that deconstruction mirrors construction ?

Personally `{ renamed_bar = bar }` makes more sense to me, since assignment normally works in that direction?

**Comment by ocharles** [comment]
Admittedly this is pretty cool, as one could imagine the RHS to be an
arbitrary expression! Like let { distance = sqrt (x*x + y*y) } = ...

**Comment by Nadrieril** [comment]
I would find that very surprising, and it's easy to do with
```dhall
let {x, y} = ...
let distance = sqrt (x*x + y*y)
in ...
```
I do agree that out of context `{ bar = renamed_bar }` would have a good chance of being understood the wrong way around. The idea behind `{ bar = renamed_bar }` instead of `{ renamed_bar = bar }` is that in pattern-matching, one would be able to write something like:
```
match record {
    { bar = Some 0, .. } => ...,
    { bar = None _, foo = { baz = 42 } } => ...,
    { bar = renamed_bar, .. } => ...,
}
```
So for consistency I think it makes more sense in this order.

**Comment by Gabriel439** [comment]
My preference is `{ bar = renamed }` when renaming a field bound in this way, for two reason:

* It's consistent with Haskell's support for pattern matching on a record
* It's the inverse of record construction (as @Nadrieril noted)

Assuming that we support renaming on deconstruction, then I don't think we need a `(baz)` syntax to ignore a field.  We can just do:

```dhall
{ foo, baz = _ }
```

... to expect and ignore a field named `baz`.  Yes, it will technically bring a variable named `_` into a scope, but in practice users aren't going to name their other variables `_` (and it will not interfere with α-normalization)

**Comment by sjakobi** [comment]
Given

```dhall
let _ = 2

let { bar = _, foo = _ } = { foo = 0, bar = 1 }

in [ _, _@1, _@2 ]
```

That should desugar to

```dhall
let _ = 2

let _ = { foo = 0, bar = 1 }.bar

let _ = { foo = 0, bar = 1 }.foo

in [ _, _@1, _@2 ]
```

and normalize to

    [ 0, 1, 2 ]

, right?

Although I guess that it would be better not to duplicate the record on the RHS for performance reasons…

---

Regarding opting out of the exhaustiveness check, is everyone ok with using `..` like with Haskell's [`RecordWildCards` extension](https://downloads.haskell.org/~ghc/latest/docs/html/users_guide/glasgow_exts.html#record-wildcards)? E.g. 

```dhall
let { map, .. } = Prelude.List
```
---

Should

```dhall
let { .. } = x
```

be legal?

---

Regarding @Gabriel439's [suggestion](https://github.com/dhall-lang/dhall-lang/issues/74#issuecomment-359087171) of supporting the same syntax in lambdas, is it ok to address that in a second step, after standardizing the record-unpacking for `let`s?

**Comment by sjakobi** [comment]
> Although I guess that it would be better not to duplicate the record on the RHS for performance reasons…

I guess a related question is what

```dhall
\(x : { foo : Text, bar : Text }) in let { foo, bar } = x in foo ++ bar
```

should normalize to:

```dhall
\(x : { foo : Text, bar : Text }) in x.foo ++ x.bar
```

or is it already in normal form?

**Comment by sjakobi** [comment]
Should `..` always appear in the last position, or is e.g.

```dhall
let { .., foo } = x
```

also ok?

**Comment by Gabriel439** [comment]
@sjakobi: Yes, that's my understanding of how the `_` example would desugar.

I also think that `let { .. } = x` should be legal, since I prefer to always support the empty case when possible for consistency

I also agree that we should standardize this in steps rather than trying to standardize the entire thing at once.

I think it's worth noting here that both Nix and Haskell support a `...`/`..` when pattern matching on records, but they mean different things.

In Nix, `...` means to ignore other fields (similar to what is being proposed here to avoid the exhaustivity check).  For Haskell's `RecordWildCards` extension, `..` means to bring all other fields into scope unqualified (similar to what was proposed in #863).

The reason I'm mentioning this is that I believe that we should actually reserve two separate symbols for dealing with "the rest of the fields" in a pattern match on a record:

* One symbol for specifying to ignore the other fields but don't bring them into scope (i.e. what Nix does)
* One symbol for specifying to bring the other fields into scope (i.e. what Haskell does)

Note that we don't necessarily need to standardize support for the latter symbol (again: some people feel pretty strongly against open imports), but I would like us to have a symbol prepared for that purpose so that if we ever change our minds and want to add open imports to the language we have a smooth migration path that builds upon this feature. (cc: @SiriusStarr)

What I will propose is:

* To ignore the rest of the fields, use `_` at the end of the pattern match, like this:

  ```dhall
  let { foo, bar, _ } = someRecord
  ```

* To bring the remaining fields into scope, use `...` at the end of the pattern match, like this:

  ```dhall
  let { foo, bar, ... } = someRecord
  ```

Given the opposition to open imports, then we only standardize the former for now, but then we know that if we ever want to introduce open imports we can then subsequently standardize the latter and a completely open import would look like this:

```dhall
let { ... } = someRecord
```

**Comment by sjakobi** [comment]
I'm fine with preparing a way forward for standardizing open imports, although I _do_ have a lot of reservations regarding such a feature.

> * To ignore the rest of the fields, use `_` at the end of the pattern match, like this:
>   ```haskell
>   let { foo, bar, _ } = someRecord
>   ```

I think that would cause confusion when `someRecord` actually has a `_` field. I'm more in favor of `..`.

If we do use `..` to ignore fields, I think `...` would be a bit too similar and could cause confusion. I think `*` could be a better fit, e.g.

    let { * } = someRecord

I believe @SiriusStarr had suggested using `*` before.

**Comment by Gabriel439** [comment]
@sjakobi: Yeah, I like the idea of using `...` to ignore fields and reserving `*` for open imports

**Comment by sjakobi** [comment]
Hm, I meant to suggest `..` (two dots) for ignoring fields. But I'm also fine with `...`. :)

What does everyone else think?

**Comment by Gabriel439** [comment]
@sjakobi: Personally, I prefer three dots since that's the standard for ellipsis.  Haskell is the only place I've seen use two dots for an ellipsis.  Also, we specifically want to convey that our ellipsis behaves like Nix's (ignore fields, using three dots) rather than Haskell's (bring fields unqualified into scope, using two dots)

**Comment by Nadrieril** [comment]
For the record, Rust also uses two dots (`..`) for ellipsis. So I'm more used to seeing that, but I have no strong opinion on the number of dots otherwise.

**Comment by Nadrieril** [comment]
Actually, I just realized we use `...` in the meta-language of the spec, e.g. we often write `{ x: T, xs... }`. So using `...` on the language too might make things confusing. I'd rather we go with `..`

**Comment by sjakobi** [comment]
Regarding normalization, desugaring to field selections would allow us to benefit from the many operator interactions with `//` and `/\` (see https://github.com/dhall-lang/dhall-lang/blob/master/standard/beta-normalization.md#records), which seems nice. :)

For example

    \(x : { foo : Text, bar : Text }) -> let { foo, bar } = x // { bar = "bla" } in foo ++ bar

could be normalized to:

    \(x : { foo : Text, bar : Text }) -> "${x.foo}bla"

**Comment by SiriusStarr** [comment]
> I also think that `let { .. } = x` should be legal, since I prefer to always support the empty case when possible for consistency

I agree legal, but maybe a lint warning (like for unused `let` bindings)?

I agree that `..` and `...` are too similar to be used here with two different meanings.  I can 100% guarantee that I'd mess up and use the wrong one, haha.

> I think that would cause confusion when someRecord actually has a `_` field. I'm more in favor of `..`

This is a good point; the symbol used should probably be one that is not a valid field name.  `{_ = 0}` is valid Dhall, but `*`, `..`, and `...` are not valid field names (without backticks) and thus are good candidates.

I honestly don't feel strongly about what is used for (/reserved for) "open" vs what is used for exhaustiveness, other than the aforementioned avoiding both `..` and `...` together.  Both get used pretty commonly for open imports (`*` in Python, `..` in Elm/Haskell/etc.), so I'm not sure there's a lot of inertia in either direction.

**Comment by sjakobi** [comment]
I like the argument that `...` is the standard symbol for ellipsis in English and many other natural languages.

> we use `...` in the meta-language of the spec, e.g. we often write `{ x: T, xs... }`. So using `...` on the language too might make things confusing. I'd rather we go with `..`

I don't expect this to be a big problem. If necessary, we could just pick a different symbol to use for the `xs…` patterns in the spec. e.g. `*xs`, similarly to the grammar.

---

> > I also think that `let { .. } = x` should be legal, since I prefer to always support the empty case when possible for consistency
> 
> I agree legal, but maybe a lint warning (like for unused `let` bindings)?

+1

---

What about unpacking `{=}` though?

IMHO

```dhall
let { ... } = {=}
```

should be an error, since "the catch-all (`...`) doesn't match any fields", just like

```dhall
let { x, ... } = { x = 42 }
```

will be an error.

Does

```dhall
let { } = {=}
```

look ok? I don't expect it to come up in practice.

---

I notice that the renaming syntax with `=` could be a bit confusing sometimes because it looks just like a record literal. For example, it's not very obvious whether this creates a binding for `x0` or a binding for `x1`:

```dhall
let { x = x1 } = { x = x0 }
```

With `as` it would be a bit clearer IMHO:

```dhall
let { x as x1 } = { x = x0 }
```

---

We'll also need a name for this construction. How about "unpacking `let`"?

"Multi-`let`" seems also like a good name, except the Haskell implementation already uses that name in the context of `let`-blocks (`let x = ... let y = ... in ...`).

**Comment by sjakobi** [comment]
And another concern regarding selective field ignoring:

```dhall
let { x, y = _ } = ...
```

will be slower to typecheck than

```dhall
let { x, (y) } = ...
```

since the binding for `_` needs to be added to the context – which we wouldn't do in the latter case.

I think this could make a real difference in code that uses that feature heavily.

Variable lookups would become more expensive too, especially when contexts are implemented as simple linked lists, like in the Haskell implementation.

**Comment by sjakobi** [comment]
> And another concern regarding selective field ignoring:

I guess a cheaper field ignoring mechanism could be added later, if necessary.

**Comment by Nadrieril** [comment]
I don't feel like that kind of small performance concern should have priority over clarity and uniformity of the language

**Comment by SiriusStarr** [comment]
> I guess a cheaper field ignoring mechanism could be added later, if necessary.

If `_` were ever added as a language-wide ignoring mechanism then we'd get this for free here.

**Comment by Gabriel439** [comment]
@sjakobi: I think this will progress more quickly if we standardize a small subset of this at a time.  I think the first step is standardizing exhaustive pattern match with `...` to ignore unused fields and don't worry about renaming or ignoring fields for the first change.

**Comment by sjakobi** [comment]
> I don't feel like that kind of small performance concern should have priority over clarity and uniformity of the language

IMHO `{ x, (y) }` will be much clearer than `{ x, y = _ }` as long as `_` is used so widely in contexts where it doesn't imply that the variable should be ignored (e.g. cached imports).

---

Here are a few more questions regarding the first step (no renaming, no selective field ignoring):

1. Is it ok if we _don't_ support leading commas for now (for consistency with [field projections](https://github.com/dhall-lang/dhall-lang/issues/814))? E.g. `let { , x } = y in z` shouldn't parse.

1. I'd like to allow `…` as an alternative to the ASCII `...`, mostly because it's 2 columns shorter. Other opinions on that?

1. Do we agree that  `let { ..., ... } = x in y` should parse but be a type error? Ergo `...` can appear at most once.

1. Regarding the position of the ellipsis, is `{ ..., x }` ok? `dhall format` should turn that into `{ x, ... }`.

1. I suggest to call this feature "record-let". Other suggestions?

**Comment by Nadrieril** [comment]
>     1. Is it ok if we _don't_ support leading commas for now (for consistency with [field projections](https://github.com/dhall-lang/dhall-lang/issues/814))? E.g. `let { , x } = y in z` shouldn't parse.

Sure, we can update both if we want to allow leading/trailing commas there.

>     2. I'd like to allow `…` as an alternative to the ASCII `...`, mostly because it's 2 columns shorter. Other opinions on that?

Sounds good.

>     3. Do we agree that  `let { ..., ... } = x in y` should parse but be a type error? Ergo `...` can appear at most once.
> 
>     4. Regarding the position of the ellipsis, is `{ ..., x }` ok? `dhall format` should turn that into `{ x, ... }`.

I'd prefer only allowing `...` at the end and only once, because I see it as a meta-thing that doesn't really make sense as an individual entity. Put differently, if we allow `let { ..., ... } = x in y` to parse then we need `...` to be a proper entity in the binary encoding, whereas if we limit to at most one `...` at the end, then the binary encoding need only have two cases for closed record pattern `{ x, y }` and open-ended record pattern `{ x, y, ... }`. I would find the second option more natural.

>     5. I suggest to call this feature "record-let". Other suggestions?

I was thinking we could instead say that what's between the `let` and the `=` is called a pattern, and a pattern is either a variable name or a record pattern. So we'd call that "record-pattern". The benefits are that we can then easily support patterns in lambdas like `\({ x, y } : T) -> foo x y` or recursive patterns like `let { x = { y, ... }, ... } = foo in y`.
This is also a further argument for only allowing `...` once in parsing: I don't feel that `...` is itself a pattern because it doesn't make sense in other pattern positions; I see it as a syntactic part of a record-pattern, and thus should be handled in parsing.

**Comment by singpolyma** [comment]
> I honestly don't feel strongly about what is used for (/reserved for) "open" vs what is used for exhaustiveness, other than the aforementioned avoiding both .. and ... together. Both get used pretty commonly for open imports (* in Python, .. in Elm/Haskell/etc.), so I'm not sure there's a lot of inertia in either direction.

I suppose this is true, but since Dhall's syntax is much more Haskellish than Pythonish it would make sense to me to borrow from related syntax?  `let {..} = x` is always what I assumed this syntax would eventually be (to bring everything from a record into scope) because I'm already used to that.  For the same reason, though, I *do* think that `...` for ignore is maybe too close?  If the argument against `_` is that it's a valid label name (in case we want punning, etc) then we could just make it *not* a valid label name and require it be quoted to be used as a label.

**Comment by MonoidMusician** [comment]
Will this syntax also be available in foralls/lambdas?

**Comment by sjakobi** [comment]
@singpolyma Note that several people don't want to get open imports standardized at all, so it's a bit of a secondary concern right now.

> If the argument against `_` is that it's a valid label name (in case we want punning, etc) then we could just make it _not_ a valid label name and require it be quoted to be used as a label.

That would break the relatively close correspondence between record labels and identifiers that we have now.

---

@MonoidMusician 

> Will this syntax also be available in foralls/lambdas?

We can eventually re-use the same syntax for lambdas.

Right now I'm focussing on standardizing the first step that Gabriel described [above](https://github.com/dhall-lang/dhall-lang/issues/74#issuecomment-570083090).

Can you explain how you'd want to use it for foralls?

**Comment by MonoidMusician** [comment]
It would work the same way for foralls as for lambdas, e.g. `forall ({ a, b, c } : { a : Type, b : Type, c : Type }) -> (a -> b) -> (b -> c) -> (a -> c)`. That way all the methods of introducing variables into context would be consistent.

I feel like I had more comments, but I can't remember them right now …

**Comment by Profpatsch** [comment]
Bikeshedding input:

* I’d prefer `..` over `...`, simply because Haskell and Rust use it.
* I prefer `let { foo as bar } = { foo = 42 };`, over `let { foo = bar } = { foo = 42; }` because
  * There’s 3 `=` in the second one, two of them must have an expression on the right side, the other one has a new symbol name on the right side
  * I could never remember which one is the old and which the new name, `as` doesn’t have that problem.
* I like the idea of explicitely ignoring fields to keep exhaustiveness-checking
  * The syntax idea with parens `{ foo, (bar) }` is not good though, because `()` should be reserved for grouping of expressions (and the eye parses it as that).
  * The syntax should be `{ foo, bar = _ }` and `_` the special “throw value away” indicator.

Non-bikeshedding input:

* I’m strongly against an open import akin to nix’s `with`, [see my comment on the other issue](https://github.com/dhall-lang/dhall-lang/issues/863#issuecomment-575953431).

**Comment by sjakobi** [comment]
Seems like we're pretty divided about the question of `..` or `...` for "ignore the remaining fields".

To quantify our preferences, please react on this comment

* with :rocket: if you prefer `..`
* with :laughing:  if you prefer `...` and its Unicode version `…`
* with :eyes: if you want something else (and specify what you want in a comment)

The outcome will be non-binding of course – it's just to inform the final decision.

(If you think that this vote is a bad idea, please also say so in a comment)

**Comment by Profpatsch** [comment]
> * and its Unicode version `…`

Oh right, I didn’t think about that. In that case I’m all in favor of `...`.

**Comment by singpolyma** [comment]
I have concerns about `...` looking like `..`, but I do buy that `_` might be problematic so if others are happy, I'm happy.

**Comment by jvanbruegge** [comment]
Sorry if I missed this, but in addition to using `...` to ignore fields, I would really like to be able use `...<ident>` to collect all unused field in a new record. For example:
```dhall
let myRec = { a = 1, b = 2, c = 3 }

let { a, ...rec } = myRec

let example1 = assert : a === 1
let example2 = assert : rec === { b = 2, c = 3 }
```

**Comment by sjakobi** [comment]
@jvanbruegge Interesting! In what kind of situations would you want to use that?

**Comment by jvanbruegge** [comment]
for example when having a function that takes a record as argument when most of the content is just for a second function you want to call. You could splice off the fields you want and give the rest to next call

**Comment by sjakobi** [comment]
I'd love to see a more concrete motivating example! 

It feels a bit like a "negated" variant of field projection. Given a record `r` of type

    { a : Bool, b : Text, c : Natural }

where you want to keep only the `a` and `b` fields, you could write either

    r.{ a, b }

or

    let { c, ...r2 } = r in r2

I think you can do something similar in Python, but I'd think that many of the Python use cases wouldn't translate so well to Dhall due to the stricter type system.

If we want to support this, we could consider re-using it for the "ignore the remaining fields" thing, with

    let { a, ..._ } = { a = 1, b = 2 } in a

I'd argue that we should still disallow it when no fields are left. E.g. this should be illegal:

    let { a, ..._ } = { a = 1 } in a

**Comment by ari-becker** [comment]
@sjakobi it would be helpful for us.

We have a pattern where we have `dhall-kubernetes` objects organized for export:

```dhall
let imports = ./imports.dhall
let Prelude = imports.Prelude
let Kubernetes = imports.Kubernetes
let foo = { service = ./foo/service.dhall , deployment = ./foo/deployment.dhall , objects = [ "service" , "deployment" ] }
let bar = { service = ./bar/service.dhall , deployment = ./bar/deployment.dhall , objects = [ "service" , "deployment" ] }
let prependPrefix = \(prefix: Text) -> \(objects: Text) -> Prelude.List.map Text Text (\(object: Text) -> "${prefix}.${object}") objects
in
{ foo = foo.{service, deployment}
, bar = bar.{service, deployment}
, objects = prependPrefix "foo" foo.objects # prependPrefix "bar" bar.objects
} : { foo : { service : Kubernetes.Service.Type, deployment: Kubernetes.Deployment.Type } 
, bar : { service : Kubernetes.Service.Type, deployment: Kubernetes.Deployment.Type }
, objects : List Text
} 
```

The idea being to run `dhall-to-json` on the whole lot and then use `jq` to iterate over `objects` so that they can be passed to `kubectl apply -f`.

The issue we currently have is in the `foo.{service.deployment}` part - adding manifests to `foo` forces changes in the exporting record (it's not a big deal but whatever). 
Being able to write `let { objects, foo } = ./foo.dhall let exported-objects.foo = objects` solves that nicely.

**Comment by Profpatsch** [comment]
I’m noticing that in order for punning to be nice, we need record punning in function arguments,
`\({ a : Text, b : Int })` should put `a` and `b` in scope.

Right now you have to do 
```
\(foo : { a : Text, b : Int }) 
  -> { a = foo.a
      , b = foo.b
```

which means you cannot pun.

**Comment by Gabriel439** [comment]
@Profpatsch: I agree that is something we should get to, but I want to note that once you have `let`-related punning then you get halfway there by permitting:

```dhall
\(foo : { a : Text, b : Int }) -> let { a, b } = foo in …
```

Indeed, support for punning function arguments could be standardized in such a way that it desugars to `let` punning

**Comment by Gabriel439** [comment]
There is a work-in-progress PR for this here: https://github.com/dhall-lang/dhall-lang/pull/1211
